README.md
# Overview
This example illustrates how to propagate custom events from the hardware to MicroUI. MicroUI already supports generic events like touch, button and joystick.
Hardware events (e.g. sensors movement) can be used for specific display action upon occurrence.

See the Device Developer's Guide at section **Inputs** (13.5) for more information.

# Usage
1. Create a platform with at least the UI module.
2. Before building this platform, some configuration is required. In the **xxx-configuration project** in the **microui/microui.xml** file, add the following xml element as the first eventgenerator.

        <eventgenerator name="CUSTOM" class="com.microej.example.customevent.CustomEventGenerator"/>

3. Build your platform.
4. In the [/com.microej.example.customevent/src/main/c](com.microej.example.customevent/src/main/c) folder
	1. Choose the file corresponding to the board/RTOS used 
	2. Replace the application entry point in the /xxx-bsp project by the file selected in the previous step
5. In MicroEJ, create a launch configuration for the Java application
	1. Set the name to "CustomEvent_Build"
	2. In the **Main** tab
		1. Set the **Project** field to "com.microej.example.customevent"
		2. Set the **Main type** field  to "CustomEventExample"
	3. In the **Execution** tab
		1. In the **Target** frame, set the JPF to the one built in step #3
		2. In the **Execution** frame, select the **Execute on EmbJPF** radio button
6. Run the launch configuration.
7. Compile, link and flash the xxx-bsp project.

# Requirements
- JRE 7 (or later) x86.
- MicroEJ 4.0 or later.
- Java platform with (at least) EDC-1.2, MICROUI-2.0.

# Project structure
- src/main/java
	- Java sources
- src/main/c
	- Entry points (main.c) examples for the given board/RTOS that generates custom events.

# Dependencies
_All dependencies are retrieved transitively by Ivy resolver_.

# Source
N/A

# Restrictions
None.


CHANGELOG.md
<!--
	Markdown
-->
<!--
Changelog template:

## Revision (YYYY-mm-dd)
Features:
  - List here the new features.
  
Bugfixes:
  - List here the bug fixes.
-->
## 1.1.1 (2017-08-16)
BugFixes:
  - Fix main
  

## 1.1.0 (2017-01-20)
Features:
  - Updated for MicroEj 4.0.
  
## 1.0.0 (2015-12-01)
Features:
  - Initial revision.



BSP_entry_points/STM32F429IDISCO_RTX_main.c
/*
 * C
 *
 * Copyright 2014 IS2T. All rights reserved.
 * For demonstration purpose only.
 * IS2T PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 */

/* Includes ------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include "sni.h"
#include "microjvm_main.h"
#include <RTL.h>

#include "stm32f4xx.h"
#include "stm32f429i_discovery_sdram.h"

#include <LLINPUT.h> // API used to send events to MicroUI.
#include <microui_constants.h> // Constants generated MicroUI initialization. It contains the id of our custom event generator.

/* Defines -------------------------------------------------------------------*/

#define MICROJVM_STACK_SIZE 2048

/* Globals -------------------------------------------------------------------*/

static U64 javaTaskStack[(MICROJVM_STACK_SIZE)/8];
 
/* Private functions ---------------------------------------------------------*/

__task void eventTask (void)
{
	for(;;)
	{
		os_dly_wait(200);
		int event = rand() % 100;
		// Sends the given event to our custom event generator.
		LLINPUT_sendEvent(MICROUI_EVENTGEN_CUSTOM, event);
	}
} 

__task void javaTask (void)
{
	os_tsk_create(eventTask, 0);
	microjvm_main();
	os_tsk_delete_self();
} 

/* Public functions ----------------------------------------------------------*/

int main(void)
{
	printf("START\n");
	
	SDRAM_Init();
	
	// start the main task
	os_sys_init_user(javaTask, 10, &javaTaskStack, sizeof(javaTaskStack));
	
	printf("END\n");
}


BSP_entry_points/STM32F429IDISCO_FreeRTOS_main.c
/*
 * C
 *
 * Copyright 2013-2015 IS2T. All rights reserved.
 * For demonstration purpose only.
 * IS2T PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 */

/* Includes ------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include "sni.h"
#include "LLMJVM.h"
#include "FreeRTOS.h"
#include "task.h"
#include "microjvm_main.h"
#include "stm32f4xx.h"
#include "stm32f429i_discovery_sdram.h"

#include <LLINPUT.h> // API used to send events to MicroUI.
#include <microui_constants.h> // Constants generated MicroUI initialization. It contains the id of our custom event generator.
/* Defines -------------------------------------------------------------------*/

#define MICROJVM_STACK_SIZE 4096
#define JAVA_TASK_PRIORITY      ( 3 ) /** Should be > tskIDLE_PRIORITY & < configTIMER_TASK_PRIORITY */
#define JAVA_TASK_STACK_SIZE     MICROJVM_STACK_SIZE/4

#define EVENT_STACK_SIZE 512
#define EVENT_TASK_PRIORITY      ( 4 ) /** Should be > JAVA_TASK_PRIORITY  & < configTIMER_TASK_PRIORITY */

/* Private functions ---------------------------------------------------------*/

void xJavaTaskFunction(void * pvParameters)
{
	microjvm_main();
	vTaskDelete( xTaskGetCurrentTaskHandle() );
}

void xEventTaskFunction(void * pvParameters)
{
	for(;;)
		{
			vTaskDelay(2000);
			int event = rand() % 100;
			// Sends the given event to our custom event generator.
			LLINPUT_sendEvent(MICROUI_EVENTGEN_CUSTOM, event);
		}
}

/* Public functions ----------------------------------------------------------*/

/*
 * Generic main function
 */
int main(void)
{
	printf("START\n");

	SDRAM_Init();
	
    // Enable fault handlers
    SCB->SHCSR |= SCB_SHCSR_MEMFAULTENA_Msk;
    SCB->SHCSR |= SCB_SHCSR_BUSFAULTENA_Msk;
    SCB->SHCSR |= SCB_SHCSR_USGFAULTENA_Msk;

	// start the main task
	xTaskCreate( xJavaTaskFunction, NULL, JAVA_TASK_STACK_SIZE, NULL, JAVA_TASK_PRIORITY, NULL );
	
	// start the event task.
	xTaskCreate( xEventTaskFunction, NULL, EVENT_STACK_SIZE, NULL, EVENT_TASK_PRIORITY, NULL );

	vTaskStartScheduler();

	printf("END\n");
}



src/main/java/com/microej/example/customevent/CustomEventGenerator.java
/*
 * Java
 *
 * Copyright 2015 IS2T. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be found at http://www.is2t.com/open-source-bsd-license/.
 */
package com.microej.example.customevent;

import ej.microui.event.Event;
import ej.microui.event.generator.GenericEventGenerator;

/**
 * Event generator used to received the custom events.
 */
public class CustomEventGenerator extends GenericEventGenerator {

	// the value matches the one generated by MicroEJ for the custom event generator
	// in the microui.xml file when building the platform
	public static final int CUSTOM = 0x00;

	@Override
	protected void eventReceived(int event) {
		event = Event.buildEvent(getEventType(), this, event);
		getEventHandler().handleEvent(event);
	}

	@Override
	public void setProperty(String name, String value) {
		//no property is to be associated with our event generator
	}

	@Override
	protected void eventsReceived(int[] events) {
		for (int event : events) {
			eventReceived(event);
		}
	}

	@Override
	public int getEventType() {
		return CUSTOM;
	}
	

}


src/main/java/com/microej/example/customevent/CustomEventDisplayable.java
/*
 * Java
 *
 * Copyright 2015 IS2T. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be found at http://www.is2t.com/open-source-bsd-license/.
 */
package com.microej.example.customevent;

import ej.microui.display.Colors;
import ej.microui.display.Display;
import ej.microui.display.Displayable;
import ej.microui.display.GraphicsContext;
import ej.microui.event.Event;
import ej.microui.util.EventHandler;

/**
 * This displayable allows to show the reception of the custom events.
 */
public class CustomEventDisplayable extends Displayable implements EventHandler {

	private static final int BACKGROUND_COLOR = Colors.NAVY;
	private static final int FOREGROUND_COLOR = Colors.WHITE;

	private final int displayWidth;
	private final int displayHeight;
	private int lastCustomEventDataReceived;
	private boolean hasReceivedEvent;
	private int eventGeneratorId;

	/**
	 * Creates a new custom event displayable for the given display.
	 *
	 * @param display
	 *            the display used by the displayable.
	 */
	public CustomEventDisplayable(Display display) {
		super(display);
		this.displayWidth = display.getWidth();
		this.displayHeight = display.getHeight();
		this.hasReceivedEvent = false;
	}

	@Override
	public void paint(GraphicsContext g) {
		int displayWidth = this.displayWidth;
		int displayHeight = this.displayHeight;

		// Fills the background.
		g.setColor(BACKGROUND_COLOR);
		g.fillRect(0, 0, displayWidth, displayHeight);

		// Draws the last custom event data received.
		String eventText;
		if (this.hasReceivedEvent) {
			eventText = "Event : data => " + this.lastCustomEventDataReceived + " , generator id => " + this.eventGeneratorId;
		} else {
			eventText = "No event";
		}
		g.setColor(FOREGROUND_COLOR);
		g.drawString(eventText, displayWidth / 2, displayHeight / 2, GraphicsContext.HCENTER | GraphicsContext.VCENTER);
	}

	@Override
	public EventHandler getController() {
		return this;
	}

	@Override
	public boolean handleEvent(int event) {
		boolean eventProcessed = false;

		// Filter on the event type to retrieve the custom events.
		if (Event.getType(event) == CustomEventGenerator.CUSTOM) {
			if (!this.hasReceivedEvent) {
				this.hasReceivedEvent = true;
			}
			this.lastCustomEventDataReceived = Event.getData(event);

			this.eventGeneratorId = Event.getGeneratorID(event);
			eventProcessed = true;
			repaint();
		}
		
		if (Event.getType(event) == Event.POINTER) {
			System.out.println("CustomEventDisplayable.handleEvent() Pointer event ID : " + Event.getGeneratorID(event));
		}
		return eventProcessed;
	}

}


src/main/c/STM32F746DISCO_FreeRTOS_main.c
/**
 ******************************************************************************
 * @file    BSP/Src/main.c
 * @author  MCD Application Team
 * @version V1.1.0RC2
 * @date    29-May-2015
 * @brief   This example code shows how to use the STM32746G Discovery BSP Drivers
 ******************************************************************************
 * @attention
 *
 * <h2><center>&copy; COPYRIGHT(c) 2015 STMicroelectronics</center></h2>
 * Copyright 2014-2016 IS2T. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *   1. Redistributions of source code must retain the above copyright notice,
 *      this list of conditions and the following disclaimer.
 *   2. Redistributions in binary form must reproduce the above copyright notice,
 *      this list of conditions and the following disclaimer in the documentation
 *      and/or other materials provided with the distribution.
 *   3. Neither the name of STMicroelectronics nor the names of its contributors
 *      may be used to endorse or promote products derived from this software
 *      without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 ******************************************************************************
 */

/* Includes ------------------------------------------------------------------*/
#include "microjvm_main.h"
#include "FreeRTOS.h"
#include "task.h"
#include "stm32746g_discovery.h"
#include "stm32746g_discovery_qspi.h"
#include <string.h>
#include <stdlib.h>
#include "cpuload.h"

#include <LLINPUT.h> // API used to send events to MicroUI.
#include <microui_constants.h> // Constants generated MicroUI initialization. It contains the id of our custom event generator.

/** @addtogroup STM32F7xx_HAL_Examples
 * @{
 */

/** @addtogroup BSP
 * @{
 */

/* Private typedef -----------------------------------------------------------*/
/* Private define ------------------------------------------------------------*/
#define MICROJVM_STACK_SIZE (12 * 1024)
#define JAVA_TASK_PRIORITY      ( 11 ) /** Should be > tskIDLE_PRIORITY & < configTIMER_TASK_PRIORITY */
#define JAVA_TASK_STACK_SIZE     MICROJVM_STACK_SIZE/4

#define EVENT_STACK_SIZE 512
#define EVENT_TASK_PRIORITY      ( 4 ) /** Should be > JAVA_TASK_PRIORITY  & < configTIMER_TASK_PRIORITY */

/* Private macro -------------------------------------------------------------*/
/* Private variables ---------------------------------------------------------*/

/* Global extern variables ---------------------------------------------------*/

/* Private function prototypes -----------------------------------------------*/
static void SystemClock_Config(void);
static void CPU_CACHE_ENABLE(void);
static void MPU_Config(void);

/* Private functions ---------------------------------------------------------*/

static void init(void)
{
	// Enable fault handlers
	SCB->SHCSR |= SCB_SHCSR_MEMFAULTENA_Msk;
	SCB->SHCSR |= SCB_SHCSR_BUSFAULTENA_Msk;
	SCB->SHCSR |= SCB_SHCSR_USGFAULTENA_Msk;

	/* Configure the MPU attributes as Write Through */
	MPU_Config();

	/* Enable the CPU Cache */
	CPU_CACHE_ENABLE();

	/* STM32F7xx HAL library initialization:
       - Configure the Flash prefetch, instruction and Data caches
       - Configure the Systick to generate an interrupt each 1 msec
       - Set NVIC Group Priority to 4
       - Global MSP (MCU Support Package) initialization
	 */
	HAL_Init();

	/* Configure the system clock to 200 Mhz */
	SystemClock_Config();

	/* Configure the Quad SPI */
    BSP_QSPI_Init();
    BSP_QSPI_MemoryMappedMode();
    HAL_NVIC_DisableIRQ(QUADSPI_IRQn);
}

__weak void LLCOMM_stack_initialize(void)
{
	// does nothing by default, overrided when ECOM-COMM is used
}

void xJavaTaskFunction(void * pvParameters)
{
	cpuload_init();
	LLCOMM_stack_initialize();
	microjvm_main();
	vTaskDelete( xTaskGetCurrentTaskHandle() );
}

void xEventTaskFunction(void * pvParameters)
{
	for(;;)
		{
			vTaskDelay(2000);
			int eventData = rand() % 100;
			// Sends the given event to our custom event generator.
			LLINPUT_sendEvent(MICROUI_EVENTGEN_CUSTOM, eventData);
		}
}

/**
 * @brief  Main program
 * @param  None
 * @retval None
 */
int main(void)
{
	init();
	printf("start\n");

	// start the main task
	xTaskCreate( xJavaTaskFunction, "MicroJvm", JAVA_TASK_STACK_SIZE, NULL, JAVA_TASK_PRIORITY, NULL );

	// start the event task.
	xTaskCreate( xEventTaskFunction, "CustomEventsGenerator", EVENT_STACK_SIZE, NULL, EVENT_TASK_PRIORITY, NULL );

	vTaskStartScheduler();	
}

void platform_main(void)
{
    main();
}

/**
  * @brief  System Clock Configuration
  *         The system Clock is configured as follow :
  *            System Clock source            = PLL (HSE)
  *            SYSCLK(Hz)                     = 200000000
  *            HCLK(Hz)                       = 200000000
  *            AHB Prescaler                  = 1
  *            APB1 Prescaler                 = 4
  *            APB2 Prescaler                 = 2
  *            HSE Frequency(Hz)              = 25000000
  *            PLL_M                          = 25
  *            PLL_N                          = 400
  *            PLL_P                          = 2
  *            PLL_Q                          = 8
  *            VDD(V)                         = 3.3
  *            Main regulator output voltage  = Scale1 mode
  *            Flash Latency(WS)              = 5
  * @param  None
  * @retval None
  */
static void SystemClock_Config(void)
{
  RCC_ClkInitTypeDef RCC_ClkInitStruct;
  RCC_OscInitTypeDef RCC_OscInitStruct;
  HAL_StatusTypeDef ret = HAL_OK;
  
  /* Enable Power Control clock */
  __HAL_RCC_PWR_CLK_ENABLE();

  /* The voltage scaling allows optimizing the power consumption when the device is
     clocked below the maximum system frequency, to update the voltage scaling value 
     regarding system frequency refer to product datasheet.  */
  __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE1);

  /* Enable HSE Oscillator and activate PLL with HSE as source */
  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSE;
  RCC_OscInitStruct.HSEState = RCC_HSE_ON;
  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
  RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSE;
  RCC_OscInitStruct.PLL.PLLM = 25;
  RCC_OscInitStruct.PLL.PLLN = 400;
  RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
  RCC_OscInitStruct.PLL.PLLQ = 8;
  
  ret = HAL_RCC_OscConfig(&RCC_OscInitStruct);
  if(ret != HAL_OK)
  {
    while(1) { ; }
  }

  /* activate the OverDrive to reach the 200 Mhz Frequency */
  ret = HAL_PWREx_ActivateOverDrive();
  if(ret != HAL_OK)
  {
    while(1) { ; }
  }

  /* Select PLL as system clock source and configure the HCLK, PCLK1 and PCLK2 clocks dividers */
  RCC_ClkInitStruct.ClockType = (RCC_CLOCKTYPE_SYSCLK | RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2);
  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
  RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
  RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV4;
  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV2;

  ret = HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_6);
  if(ret != HAL_OK)
  {
    while(1) { ; }
  }
}

/**
 * @brief  CPU L1-Cache enable.
 *         Invalidate Data cache before enabling
 *         Enable Data & Instruction Cache
 * @param  None
 * @retval None
 */
static void CPU_CACHE_ENABLE(void)
{
	(*(uint32_t *) 0xE000ED94) &= ~0x5;
	(*(uint32_t *) 0xE000ED98) = 0x0; //MPU->RNR
	(*(uint32_t *) 0xE000ED9C) = 0x20010000 |1<<4; //MPU->RBAR
	(*(uint32_t *) 0xE000EDA0) = 0<<28 | 3 <<24 | 0<<19 | 0<<18 | 1<<17 | 0<<16 | 0<<8 | 30<<1 | 1<<0 ; //MPU->RASE  WT
	(*(uint32_t *) 0xE000ED94) = 0x5;

	/* Invalidate I-Cache : ICIALLU register*/
	SCB_InvalidateICache();

	/* Enable branch prediction */
	SCB->CCR |= (1 <<18);
	__DSB();

	/* Enable I-Cache */
	SCB_EnableICache();

	/* Enable D-Cache */
	SCB_InvalidateDCache();
	SCB_EnableDCache();
}

/**
 * @brief  Configure the MPU attributes as Write Through for SRAM1/2.
 * @note   The Base Address is 0x20010000 since this memory interface is the AXI.
 *         The Region Size is 256KB, it is related to SRAM1 and SRAM2  memory size.
 * @param  None
 * @retval None
 */
static void MPU_Config(void)
{
	MPU_Region_InitTypeDef MPU_InitStruct;

	/* Disable the MPU */
	HAL_MPU_Disable();

	/*
	 * WT == write-through
	 * 	C == 1
	 * 	B == 0
	 * 	T == 00
	 *
	 * SO == strongly ordered
	 * 	C == 1
	 * 	B == 0
	 * 	T == 00
	 *
	 * 	FIXME: update linker file (bsp and standalone) when this configuration changes
	 */

	/* Configure the MPU attributes as WT for SRAM */
	MPU_InitStruct.Enable = MPU_REGION_ENABLE;
	MPU_InitStruct.BaseAddress = 0x20010000;
	MPU_InitStruct.Size = MPU_REGION_SIZE_256KB;
	MPU_InitStruct.AccessPermission = MPU_REGION_FULL_ACCESS;
	MPU_InitStruct.IsBufferable = MPU_ACCESS_NOT_BUFFERABLE;
	MPU_InitStruct.IsCacheable = MPU_ACCESS_CACHEABLE;
	MPU_InitStruct.IsShareable = MPU_ACCESS_NOT_SHAREABLE;
	MPU_InitStruct.Number = MPU_REGION_NUMBER0;
	MPU_InitStruct.TypeExtField = MPU_TEX_LEVEL0;
	MPU_InitStruct.SubRegionDisable = 0x00;
	MPU_InitStruct.DisableExec = MPU_INSTRUCTION_ACCESS_ENABLE;
	HAL_MPU_ConfigRegion(&MPU_InitStruct);

	/* Configure the MPU attributes as SO for SDRAM for display*/
	MPU_InitStruct.Enable = MPU_REGION_ENABLE;
	MPU_InitStruct.BaseAddress = 0xC0000000;
	MPU_InitStruct.Size = MPU_REGION_SIZE_4MB;
	MPU_InitStruct.AccessPermission = MPU_REGION_FULL_ACCESS;
	MPU_InitStruct.IsBufferable = MPU_ACCESS_NOT_BUFFERABLE;
	MPU_InitStruct.IsCacheable = MPU_ACCESS_NOT_CACHEABLE;
	MPU_InitStruct.IsShareable = MPU_ACCESS_NOT_SHAREABLE;
	MPU_InitStruct.Number = MPU_REGION_NUMBER1;
	MPU_InitStruct.TypeExtField = MPU_TEX_LEVEL0;
	MPU_InitStruct.SubRegionDisable = 0x00;
	MPU_InitStruct.DisableExec = MPU_INSTRUCTION_ACCESS_ENABLE;
	HAL_MPU_ConfigRegion(&MPU_InitStruct);

	/* Configure the MPU attributes as WT for SDRAM (remove 512K from display part)*/
	MPU_InitStruct.Enable = MPU_REGION_ENABLE;
	MPU_InitStruct.BaseAddress = 0xC0380000;
	MPU_InitStruct.Size = MPU_REGION_SIZE_512KB;
	MPU_InitStruct.AccessPermission = MPU_REGION_FULL_ACCESS;
	MPU_InitStruct.IsBufferable = MPU_ACCESS_NOT_BUFFERABLE;
	MPU_InitStruct.IsCacheable = MPU_ACCESS_CACHEABLE;
	MPU_InitStruct.IsShareable = MPU_ACCESS_NOT_SHAREABLE;
	MPU_InitStruct.Number = MPU_REGION_NUMBER2;
	MPU_InitStruct.TypeExtField = MPU_TEX_LEVEL0;
	MPU_InitStruct.SubRegionDisable = 0x00;
	MPU_InitStruct.DisableExec = MPU_INSTRUCTION_ACCESS_ENABLE;
	HAL_MPU_ConfigRegion(&MPU_InitStruct);

	/* Configure the MPU attributes as WT for SDRAM (remaining area)*/
	MPU_InitStruct.Enable = MPU_REGION_ENABLE;
	MPU_InitStruct.BaseAddress = 0xC0400000;
	MPU_InitStruct.Size = MPU_REGION_SIZE_4MB;
	MPU_InitStruct.AccessPermission = MPU_REGION_FULL_ACCESS;
	MPU_InitStruct.IsBufferable = MPU_ACCESS_NOT_BUFFERABLE;
	MPU_InitStruct.IsCacheable = MPU_ACCESS_CACHEABLE;
	MPU_InitStruct.IsShareable = MPU_ACCESS_NOT_SHAREABLE;
	MPU_InitStruct.Number = MPU_REGION_NUMBER3;
	MPU_InitStruct.TypeExtField = MPU_TEX_LEVEL0;
	MPU_InitStruct.SubRegionDisable = 0x00;
	MPU_InitStruct.DisableExec = MPU_INSTRUCTION_ACCESS_ENABLE;
	HAL_MPU_ConfigRegion(&MPU_InitStruct);

	/* Configure the MPU attributes as SO for QSPI (unused zone) */
	MPU_InitStruct.Enable = MPU_REGION_ENABLE;
	MPU_InitStruct.BaseAddress = 0x90000000;
	MPU_InitStruct.Size = MPU_REGION_SIZE_256MB;
	MPU_InitStruct.AccessPermission = MPU_REGION_FULL_ACCESS;
	MPU_InitStruct.IsBufferable = MPU_ACCESS_NOT_BUFFERABLE;
	MPU_InitStruct.IsCacheable = MPU_ACCESS_NOT_CACHEABLE;
	MPU_InitStruct.IsShareable = MPU_ACCESS_NOT_SHAREABLE;
	MPU_InitStruct.Number = MPU_REGION_NUMBER4;
	MPU_InitStruct.TypeExtField = MPU_TEX_LEVEL0;
	MPU_InitStruct.SubRegionDisable = 0x00;
	MPU_InitStruct.DisableExec = MPU_INSTRUCTION_ACCESS_ENABLE;
	HAL_MPU_ConfigRegion(&MPU_InitStruct);

	/* Configure the MPU attributes as WT for QSPI (used 16Mbytes) */
	MPU_InitStruct.Enable = MPU_REGION_ENABLE;
	MPU_InitStruct.BaseAddress = 0x90000000;
	MPU_InitStruct.Size = MPU_REGION_SIZE_16MB;
	MPU_InitStruct.AccessPermission = MPU_REGION_FULL_ACCESS;
	MPU_InitStruct.IsBufferable = MPU_ACCESS_NOT_BUFFERABLE;
	MPU_InitStruct.IsCacheable = MPU_ACCESS_CACHEABLE;
	MPU_InitStruct.IsShareable = MPU_ACCESS_NOT_SHAREABLE;
	MPU_InitStruct.Number = MPU_REGION_NUMBER5;
	MPU_InitStruct.TypeExtField = MPU_TEX_LEVEL0;
	MPU_InitStruct.SubRegionDisable = 0x00;
	MPU_InitStruct.DisableExec = MPU_INSTRUCTION_ACCESS_ENABLE;
	HAL_MPU_ConfigRegion(&MPU_InitStruct);

	/* Enable the MPU */
	HAL_MPU_Enable(MPU_PRIVILEGED_DEFAULT);
}

#ifdef USE_FULL_ASSERT

/**
 * @brief  Reports the name of the source file and the source line number
 *         where the assert_param error has occurred.
 * @param  file: pointer to the source file name
 * @param  line: assert_param error line source number
 * @retval None
 */
void assert_failed(uint8_t* file, uint32_t line)
{
	/* User can add his own implementation to report the file name and line number,
     ex: printf("Wrong parameters value: file %s on line %d\r\n", file, line) */

	/* Infinite loop */
	while (1)
	{
	}
}
#endif /* USE_FULL_ASSERT */

/**
 * @}
 */

/**
 * @}
 */

/************************ (C) COPYRIGHT STMicroelectronics *****END OF FILE****/
